### DISCUSSION TITLE
Make PDF read only

### DISCUSSION THREAD

**Discussion by thusharagokulnath** [py-pdf/pypdf#2627] [Q&A] [answered]
I would like to know is there any method to make PDF read only using the pyPDF extension?
I tried to read few issues related to this, but couldn't find exact answer.

**Comment by stefan6419846** [comment]
What do you mean by making it read-only? Some PDF viewers might open PDF/A documents as read-only by default.

Additionally, you might want to have a look at setting an owner password and corresponding user access permissions, for example `writer.encrypt(user_password="", owner_password="abc", permissions_flag=UserAccessPermissions(0))`.

**Reply by thusharagokulnath** [reply]
I am updating data to a editable PDF using pyPDF. After updating , the saved PDF should be read only for the users in all PDF viewers like adobe or browsers. That is the requirement

**Reply by thusharagokulnath** [reply]
writer.encrypt(user_password="", owner_password="abc", permissions_flag=UserAccessPermissions(0)) is disabling the edit.But it disabling the Signature as well. It looks editable but can not click and do the signature

**Reply by stefan6419846** [reply]
Please have a look at the different permissions you can set for PDF files inside the docs or inside the PDF specification: https://pypdf.readthedocs.io/en/stable/modules/constants.html#pypdf.constants.UserAccessPermissions

Example: Disallow everything except filling forms is doable by using the following permissions: 

```python3
from pypdf.constants import UserAccessPermissions as UAP

permissions = UAP(UAP.all() - UAP.FILL_FORM_FIELDS)
```

**Reply by thusharagokulnath** [reply]
Thank you. But where can I see, what does the constants R1, R2 ... etc means? 
Also couldn't find any UAP which allows Signature field to be editable and all other fields to be disabled

**Reply by stefan6419846** [reply]
The `R*` values are reserved positions which you can safely ignore. For an explanation of the named values, please consult the PDF specification. I am not sure whether your use case with only the signature field can be covered by user access permissions and is something you can only test - you might have to consider another approach if it does not work.

**Reply by thusharagokulnath** [reply]
Thank you for the reply. I have tried all the UAP given. Is there any approach to disable only specific fields ?

**Reply by pubpub-zz** [reply]
you should have a look at Pdf Spec to Ff  entry (applicable to all fields) 
![image](https://github.com/py-pdf/pypdf/assets/4083478/b4138801-772b-4781-96f1-417bdbc81a10)
this should make readonly the fields you are looking for

**Reply by thusharagokulnath** [reply]
Can you please provide a code snippet for this

**Reply by pubpub-zz** [reply, marked as answer]
Something like that
```
import pypdf
r = pypdf.PdfReader("resources/FormTestFromOo.pdf")
for k,v in r.get_fields():
    o = v.indirect_reference.get_object()
    if o["/FT"] != "/Sig":
        o[pypdf.generic.NameObject("/Ff")] = pypdf.generic.NumberObject(o.get("/Ff",0) | 1)
```

**Reply by thusharagokulnath** [reply]
Thank you

**Reply by pubpub-zz** [reply]
you should also have a look in the manual:
https://pypdf.readthedocs.io/en/stable/modules/PdfWriter.html#pypdf.PdfWriter.update_page_form_field_values

**Reply by Usouf** [reply]
> Something like that
> 
> ```
> import pypdf
> r = pypdf.PdfReader("resources/FormTestFromOo.pdf")
> for k,v in r.get_fields():
>     o = v.indirect_reference.get_object()
>     if o["/FT"] != "/Sig":
>         o[pypdf.generic.NameObject("/Ff")] = pypdf.generic.NumberObject(o.get("/Ff",0) | 1)
> ```

change the code to the following:
`for k,v in r.get_fields().items():`
